src/ledger/LedgerManagerImpl.cpp

```diff
@@ -388,47 +408,42 @@ LedgerManagerImpl::valueExternalized(LedgerCloseData const& ledgerData)
     {
     case LedgerManager::LM_BOOTING_STATE:
     case LedgerManager::LM_SYNCED_STATE:
-        if (mLastClosedLedger.header.ledgerSeq + 1 == ledgerData.getLedgerSeq())
+    {
+        switch (closeLedgerIf(ledgerData))
         {
-            if (mLastClosedLedger.hash ==
-                ledgerData.getTxSet()->previousLedgerHash())
-            {
-                setState(LM_SYNCED_STATE);
-                closeLedger(ledgerData);
-                CLOG(INFO, "Ledger")
-                    << "Closed ledger: " << ledgerAbbrev(mLastClosedLedger);
-            }
-            else
-            {
-                CLOG(FATAL, "Ledger") << "Network consensus for ledger "
-                                      << mLastClosedLedger.header.ledgerSeq
-                                      << " changed; this should never happen";
-                throw std::runtime_error("Network consensus inconsistency");
-            }
-        }
-        else if (ledgerData.getLedgerSeq() <=
-                 mLastClosedLedger.header.ledgerSeq)
+        case CloseLedgerIfResult::CLOSED:
         {
-            CLOG(INFO, "Ledger")
-                << "Skipping close ledger: local state is "
-                << mLastClosedLedger.header.ledgerSeq << ", more recent than "
-                << ledgerData.getLedgerSeq();
+            setState(LM_SYNCED_STATE);
         }
-        else
+        break;
+        case CloseLedgerIfResult::TOO_OLD:
+            // nothing to do
+            break;
+        case CloseLedgerIfResult::TOO_NEW:
         {
-            // Out of sync, buffer what we just heard and start catchup.
-            CLOG(INFO, "Ledger")
-                << "Lost sync, local LCL is "
-                << mLastClosedLedger.header.ledgerSeq
-                << ", network closed ledger " << ledgerData.getLedgerSeq();
-
             initializeCatchup(ledgerData);
         }
         break;
+        }
+    }
+    break;
 
     case LedgerManager::LM_CATCHING_UP_STATE:
```

Comment:
The code would be simpler and easier to understand if we only had one list of states for the state machine: right now you have nested switches and state transitions are nested too, this is not how people typically describe state machines and while it looks like name mangling, it forces to enumerate all possible states for the state machine and all transitions.

So, replace `LM_CATCHING_UP_STATE` by something like this:
```
LM_CATCHING_UP_WAITING_TRIGGER_STATE,
LM_CATCHING_UP_APPLYING_STATE,
LM_CATCHING_UP_FINALIZING_STATE
```

Reply:
I did it for two reasons:
1. adding LM_CATCHING_UP... states complicated many other switch statements which only care if  any of LM_CATCHING_UP... is on
2. at some point I hope that whole catchup code will be in CatchupManager, then it will definitely require separate state machine, which is now nested in LedgerManager state machine

Reply:
yes I agree it makes sense in the context of a refactor of that code into `CatchupManager`, I guess we have to prioritize this refactor. I created issue #1674 to track this.
I'll merge this PR now anyways then